### PR TITLE
feat: Create generic events feature

### PR DESCRIPTION
**Pull request by metal-messiah**
Create a feature that will normalize, aggregate and harvest all generic events.
---

### Related Issue(s)
NR-291171

### Testing
Feature testing to be added when real data begins flowing through the new feature

### CODE REVIEW

File: src/cdn/experimental.js [hunk 1]
```diff
@@ -17,6 +17,7 @@ import { Instrument as InstrumentErrors } from '../features/jserrors/instrument'
 import { Instrument as InstrumentXhr } from '../features/ajax/instrument'
 import { Instrument as InstrumentSessionTrace } from '../features/session_trace/instrument'
 import { Instrument as InstrumentSessionReplay } from '../features/session_replay/instrument'
+import { Instrument as InstrumentGenericEvents } from '../features/generic_events/instrument'
 // import { Instrument as InstrumentSpa } from '../features/spa/instrument'
 import { Instrument as InstrumentSoftNav } from '../features/soft_navigations/instrument'
 import { Instrument as InstrumentPageAction } from '../features/page_action/instrument'
```
Comment (patrickhousley) [inline src/cdn/experimental.js:23]: Doesn't this need to be removed?

**Review by patrickhousley — CHANGES_REQUESTED** [review]
Missing package.json updates for this feature and for soft nav. Should also update the npm tests.

File: src/cdn/experimental.js (hunk 1 above)
Comment (metal-messiah) [reply src/cdn/experimental.js:23]: If you are referring to PageActions being removed from the CDN entries, IMO that should be done with the re-write of the feature for better consistency of the changes

Comment (patrickhousley) [reply src/cdn/experimental.js:23]: I just noticed you removed it in the pro polyfills loader. Just wondering if maybe one of these is a mistake?

Comment (cwli24) [reply src/cdn/experimental.js:23]: @metal-messiah better would be to leave this out then and replace PA with GE in that PR

File: src/features/generic_events/aggregate/index.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    if (options.retry) {
+      this.currentEvents = harvestEvents
+    }
+
+    return payload
+  }
+
+  onHarvestFinished (result) {
+    if (result && result.sent && result.retry && this.currentEvents) {
+      this.events = this.events.concat(this.currentEvents)
```
Comment (cwli24) [inline src/features/generic_events/aggregate/index.js:67]: currentEvents should maintain first in order

File: src/features/generic_events/aggregate/index.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+    /** should have been provided by reporting feature -- but falls back to now if not */
+    eventAttributes.timestamp ??= this.#agentRuntime.timeKeeper.convertRelativeTimestamp(now())
+
+    this.events.push(eventAttributes)
+
+    // check if we've reached the harvest limit...
+    if (this.events.length >= this.eventsPerHarvest) {
+      this.overflow = [...this.overflow, ...this.events.splice(0, Infinity)]
+      this.harvestScheduler.runHarvest()
```
Comment (cwli24) [inline src/features/generic_events/aggregate/index.js:105]: need to align on all the individual max buffer harvest logic. Low interval harvesting is becoming pointless if everything can just call harvest when full

Comment (metal-messiah) [reply src/features/generic_events/aggregate/index.js:105]: mainly there just protect unload behavior

File: src/features/generic_events/aggregate/index.js (hunk 2 above)
Comment (metal-messiah) [reply src/features/generic_events/aggregate/index.js:67]: why does this matter?  Every event is timestamped

Comment (cwli24) [reply src/features/generic_events/aggregate/index.js:72]: just the pattern with all the other feature like trace; it's also implied when we troubleshoot and open payloads in dev tools to find the earliest at the top

File: src/cdn/experimental.js (hunk 1 above)
Comment (metal-messiah) [reply src/cdn/experimental.js:23]: good catch

File: package.json [hunk 4]
```diff
@@ -53,6 +56,9 @@
       "features/session_trace": [
         "dist/types/features/session_trace/index.d.ts"
       ],
+      "features/session_replay": [
+        "dist/types/features/session_replay/index.d.ts"
+      ],
```
Comment (patrickhousley) [inline package.json:61]: Still missing the typings entry for soft nav but thats not really part of this PR so I understand if you want to do that separately.

File: src/features/generic_events/aggregate/index.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+      if (!ins) {
+        this.blocked = true
+        deregisterDrain(this.agentIdentifier, this.featureName)
+        return
+      }
+
+      // handle page actions and other generic events here
+      this.harvestScheduler = new HarvestScheduler('ins', { onFinished: (...args) => this.onHarvestFinished(...args) }, this)
+      this.harvestScheduler.harvest.on('ins', (...args) => this.onHarvestStarted(...args))
+      // this.harvestScheduler.startTimer(this.harvestTimeSeconds, 0)
```
Comment (patrickhousley) [inline src/features/generic_events/aggregate/index.js:45]: I assume this line will be uncommented once we add a sub-feature.

File: src/features/generic_events/aggregate/index.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+    for (let key in obj) {
+      let val = obj[key]
+      if (key === 'timestamp') val = this.#agentRuntime.timeKeeper.correctAbsoluteTimestamp(val)
+      obj[key] = (val && typeof val === 'object' ? stringify(val) : val)
+    }
+
+    const eventAttributes = {
+      /** Agent-level custom attributes */
+      ...(getInfo(this.agentIdentifier).jsAttributes || {}),
```
Comment (patrickhousley) [inline src/features/generic_events/aggregate/index.js:94]: Should jsAttributes be last in the list so that it gets truncated if the size is too large?

File: src/features/generic_events/aggregate/index.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+    /** should have been provided by reporting feature -- but falls back to now if not */
+    eventAttributes.timestamp ??= this.#agentRuntime.timeKeeper.convertRelativeTimestamp(now())
+
+    this.events.push(eventAttributes)
+
+    // check if we've reached the harvest limit...
+    if (this.events.length >= this.eventsPerHarvest) {
+      this.overflow = [...this.overflow, ...this.events.splice(0, Infinity)]
+      this.harvestScheduler.runHarvest()
```
Comment (patrickhousley) [inline src/features/generic_events/aggregate/index.js:111]: We should put in a ticket to centralize the harvest buffer mechanics so that we are not re-inventing the wheel in every feature around max harvest size, etc.

**Review by patrickhousley — APPROVED** [review]
All the comments I left were informational and don't necessarily require a change here. Approving.

File: src/features/generic_events/aggregate/index.js (hunk 6 above)
Comment (metal-messiah) [reply src/features/generic_events/aggregate/index.js:94]: yes and also so that it matches the same attribute precedence

File: src/features/generic_events/aggregate/index.js (hunk 7 above)
Comment (metal-messiah) [reply src/features/generic_events/aggregate/index.js:111]: yeah, needs a middle man to buffer instead of feature by feature so we can check, but agreed
